### PR TITLE
Rzk auto-installation

### LINKED ISSUE #21: Prompt installing/updating rzk when not found in PATH
If there is no `rzk` executable found in the `PATH`, provide an option (through a prompt displayed automatically to the user) to download the latest (supported) version from GitHub releases.

This also involves storing the supported rzk version range in the extension.
Additionally, a prompt should be displayed to update when a newer (supported) version than the one in use is found on GitHub.

The priority for which binary to use should be as follows:
- The path provided in the settings (depends on #19)
- The global one on the PATH 
- The version downloaded locally for the extension

### PR DESCRIPTION
**Pull request by aabounegm**
Implements a prompt to install `rzk` (local to VS Code only) if it was not already found on `PATH` and not configured in the extension settings:
image

including a progress indicator while installing:
image

Additionally, it will prompt to update the local version (not the one from `PATH`) if updates are available on GitHub:
image

The installation process is simply fetching the latest compatible release on GitHub with an asset compatible with the current platform and then extracting it in the global storage location of our extension.

Resolves #21 

# Known limitations

Only the latest downloaded version is kept, so if we will support specifying the rzk version in the project settings in the future, we will need to possibly store more than one version.

It should also be better documented that installing `rzk` globally after having installed it through the extension will cause the global version (from `PATH`) to take precedence.

Lastly, version comparison is done by coercing the GitHub release tag to [semver](https://semver.org/), not PVP. This hopefully shouldn't cause a problem (according to the couple of manual tests I tried), but might limit how we can specify supported version ranges.

### CODE REVIEW

**Comment by fizruk** [comment]
I've removed `rzk` locally, but do not see the prompt. How to debug this?

EDIT: I forgot to do `npm i && npm run compile`, the prompt works!

**Comment by fizruk** [comment]
After installation, `rzk` does not appear to be working: no semantic highlighting and `rzk` is not available in Terminal.

EDIT: false alarm, I was testing code with unreleased features.

File: src/installRzk.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    const path = join(binFolder.fsPath, localBin[0]);
+    output.appendLine('Found local installation of rzk: ' + path);
+    await checkForUpdates(path, binFolder);
+    return;
+  }
+
+  // Ignore the returned promise to not block the rest of the code waiting for user input
+  void vscode.window
+    .showWarningMessage(
+      "Cannot find 'rzk' in PATH. Install it?",
```
Comment (fizruk) [inline src/installRzk.ts]:
```suggestion
      "Cannot find 'rzk' in PATH. Install latest version of rzk from GitHub releases?",
```

File: src/installRzk.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+      'Yes',
+      'Ignore'
+    )
+    .then(async (value) => {
+      if (value === 'Yes') {
+        await vscode.window.withProgress(
+          {
+            title: 'Installing rzk...',
```
Comment (fizruk) [inline src/installRzk.ts]: Can we add specific version being installed?

File: src/installRzk.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+      `Version ${version} of rzk is installed, but ${latestRelease.tag_name} is available. Would you like to update?`,
+      'Yes',
+      'No'
+    )
+    .then(async (value) => {
+      if (value === 'Yes') {
+        output.appendLine('Updating local rzk version');
+        await vscode.window.withProgress(
+          {
+            title: 'Updating rzk...',
```
Comment (fizruk) [inline src/installRzk.ts]: Would be nice to specify versions:
```
            title: 'Updating rzk (v0.2.0 => v0.5.1)...',
```

**Comment by fizruk** [comment]
Once this is finished and merged, we need to update documentation [here](https://github.com/fizruk/rzk/blob/develop/docs/docs/getting-started/install.md).

File: src/installRzk.ts (hunk 2 above)
Comment (aabounegm) [reply src/installRzk.ts]: We only have the version inside the `installLatestRzk` function itself. I can refactor it (perhaps tomorrow) to split into 2 functions, one to get the version and the other to install it, or I can send the `progress` object to that function instead and update the message from there (though I'm not sure how I feel about that, feels a bit untidy)

Comment (aabounegm) [reply src/installRzk.ts]: Actually, nvm that comment, that is a local function and passing `progress` actually makes sense, especially if we want to report back the intermediate steps as well.

**Comment by fizruk** [comment]
How to uninstall `rzk` which was installed by the extension?

**Comment by aabounegm** [comment]
For now it's not possible yet. It can be done manually by going to the extension's global storage folder (`%AppData%\Code\User\globalStorage\nikolaikudasovfizruk.rzk-1-experimental-highlighting` on Windows, not sure on macOS/Linux, but you can see it in the extension's logs) and deleting the contents of the `bin` folder. We can add a command to clear that folder.
However, I don't see why this would be needed since the one installed by extension has the last priority to be used and doesn't affect the rest of the system.

**Comment by fizruk** [comment]
Well, I want to be able to test installation :)

**Comment by fizruk** [comment]
> ... but you can see it in the extension's logs) and deleting the contents of the `bin` folder. We can add a command to clear that folder.

Thanks! I've been able to find the path and remove manually! For macOS this was the path:

```
/Users/%USERNAME%/Library/Application Support/Code/User/globalStorage/nikolaikudasovfizruk.rzk-1-experimental-highlighting/bin/rzk
```

It would great to at least document how to get rid of a local installation in case it is broken, or people want to revert to another version, or for testing, etc.

**Comment by aabounegm** [comment]
Alright, I added a command to delete it that is only available while in debug mode (to not crowd users' command palettes since it's mainly only useful while developing the extension).

**Comment by fizruk** [comment]
Very nice! Looks good to me :)

**Review by fizruk — APPROVED** [review]

**Pull request merged**